[JUJU-563, JUJU-730] Per runtime gc

arnodel opened the pull request:
In Lua tables / userdata can be provided with a finalizer by implementing their __gc metamethod.  This is used e.g. in file objects, so that the underlying file descriptor is released before garbage collection of the file object.

In Golua this feature is unimplemented so far, except for a special case for files.  The problem is that if we have many Golua runtimes running in the same process, we probably want to run all finalizers belonging to a given runtime when we close this runtime.

This task is about investigating a way to achieve that.  It is difficult because if a runtime keeps a list of objects with finalizers, then obviously they will never get gc’ed.  Weak references would solve the issue but there is not weak ref API in Go.

## Aims

- to make sure that when a runtime is closed, all Lua pending finalizers are run.  The associated values may not yet have been GCed by the Go runtime but this feels necessary in order to isolate runtimes
- to make sure that when new objects are created in a resource limiting runtime context with a finalizer, that this finalizer will not consume resources of other contexts when it eventually runs.  This may be done in several ways:
  * attempt to run finalizers before exiting the context - there is an understanding that objects created in such a context become finalized when the context is exited.  If the context has no more resources then that is aborted.
  * disallow registration of finalizers in the context
  * attempt to run finalizers for objects created in the context after they become unreachable, at an appropriate time such that they do not consume resources of other contexts.  This approach would be the most general but trickiest.
  * disallow creation of nested context
- to make sure that userdata that hold system resources (e.g. file descriptors) will release those resources even if they were created in a resource limiting runtime context.

## Approach

### Pools to manage values with finalizers

The approach take here also may prepare the way for supporting weak references.  There is a package called `weakref` implementing a `weakref.Pool` interface that allows the management of object finalizers.  It can tell which objects are ready to be finalized, and which objects have had a finalizer registered but not yet triggered.

The Lua runtime uses such a `Pool` to decide when to run object finalizers (they must be run synchronously with the runtime thread of execution).

Userdata in the Pool which implement the `UserDataResourceReleaser` interface are guaranteed to run their resource releasing code when they are removed from the pool, always after the lua finalizer if it exists.  This is used by file userdata to make sure file descriptors are released.

There is a gc policy associated with a runtime context which has two valid values "share" and "isolate".
- "share" means that the pool is shared with the parent context, therefore its contents are not removed when the context ends
- "isolate" means the pool is new, so when the context ends the pool is removed.  This means also that finalizing code from parent context is deferred until the current context ends.

The gc policy does not need to be set explicitly, it is automatically set to "isolate" for a context which limits resources and to "share" for other contexts.  It is currently possible to override "share" in the Go API but not in Lua.

### Disallow nested contexts

If a value with a finalizer is created in a restricted runtime context, then the finalizer code should run within this context.  However it is possible that the value will outlive the context (i.e. still be reachable in the outside context).  In that case it does not make sense to run its finalizer when exiting the context.

It seems difficult to reconcile those two contradictory constraints.  The simplest way to resolve this issue is to disallow nested runtime contexts, so that it will be the runtime itself which will be constrained.  This is the approach taken in this PR.  However it is still useful to keep the notion of runtime context as it encapsulates all concerns about resource management.

Many values in the runtime that have a finalizer may be created, so it is important to be able to finalize them before the runtime closes (e.g. if many files are created, the finalizer will be in charge of making sure their file descriptor is closed).  An implementation of `weakref.Pool` is provided that does that (`weakref.ClonePool`).  That implementation does not permit the creation of weak references but that is a limitation that already existed in Golua.

arnodel committed:
Ensure gc operates per runtime

arnodel committed:
Make sure UserData has its __gc metamethod registered

arnodel committed:
Move weakref to own internal package

arnodel committed:
Document how Golua runtime uses weak refs.

arnodel committed:
Implement isolating gc in resource constrained contexts

arnodel committed:
Merge branch 'lua5.4' into per-runtime-gc

arnodel committed:
Fix build for noquotas

arnodel committed:
Runtime.Close tries to clean all contexts

arnodel committed:
Fix noquotas runtimemanager

arnodel committed:
Try to fix gc test

arnodel committed:
Fix sorting in weakref pool

arnodel committed:
Add test for finalizers in contrained context

arnodel committed:
Remove debug print

arnodel committed:
Add sections in quotaas doc about finalizers

arnodel committed:
Merge remote-tracking branch 'canonical/lua5.4' into per-runtime-gc

arnodel committed:
Rename orderedVals to sortableVals

arnodel committed:
Refactor weakref pool to guarantee releasing resources after lua finalizers run

arnodel committed:
Relax time test

arnodel committed:
Lua tests and bugfixes

arnodel committed:
Separate quotas and non quotas tests

arnodel committed:
Run tests with safe and unsafe pool

arnodel committed:
Fix pool creation in noquotas build

arnodel committed:
Tests for UnsafePool

arnodel committed:
Tests for SafePool

arnodel committed:
Don't set finalizer for resurrected values

arnodel committed:
Do not set finalizer on non-resurrected values

arnodel committed:
Merge branch 'lua5.4' into per-runtime-gc

arnodel committed:
lock for longer

arnodel committed:
Fix doc comment

arnodel committed:
ClonePool implementation

This implementation of weakref.Pool  allows GC to happen any time
but does not support weakrefs

arnodel committed:
Test that clone pool gives no weak refs

arnodel committed:
Test that clone pool returns values in correct order

arnodel committed:
Improve naming of private types/fields in clonepool

arnodel committed:
Improve doc comments for weakref package

arnodel committed:
Allow capturing error when closing runtime

arnodel committed:
Add convenience functions for running code with quotas

SimonRichardson commented on runtime/internal/weakref/clonepool.go:
```diff
@@ -0,0 +1,225 @@
+// ClonePool is an implementation of Pool that makes every effort to let values
+// be GCed when they are only reachable via WeakRefs.  Unlike UnsafePool, it
+// doesn't rely on any undocumented properties of the Go runtime so it is safe
+// to use in any compliant Go implementation.  The downsid of this
```
```suggestion
// to use in any compliant Go implementation.  The downside of this
```

SimonRichardson commented on runtime/internal/weakref/clonepool.go:
```diff
@@ -0,0 +1,225 @@
+// finalizers. The values returned are removed from the pool and their weak refs
+// are invalidated.
+func (p *ClonePool) ExtractPendingFinalize() []Value {
+	p.mx.Lock()
+	pending := p.pendingFinalize
+	if pending == nil {
+		p.mx.Unlock()
+		return nil
+	}
+	p.pendingFinalize = nil
+	p.mx.Unlock()
```
This pattern is used in multiple cases and because of the weird locking semantics, I think extracting this to a separate function is wise. Then you can simplify this to `Lock` then `defer Unlock`.

arnodel replied:
It's used in two places, mutating a different field so personally I don't feel like it warrants factoring out.  I could add some comment explaining what's happening?

SimonRichardson commented on runtime/internal/weakref/clonepool.go:
```diff
@@ -0,0 +1,225 @@
+		p.pendingRelease = append(p.pendingRelease, c)
+	}
+
+	delete(p.cloneRegister, v.Key())
```
```suggestion
	delete(p.cloneRegister, k)
```

SimonRichardson commented on runtime/internal/weakref/defaultpool.go:
```diff
@@ -0,0 +1,8 @@
+//go:build !safepool
+
+package weakref
```
Considering the clone pool isn't actually a weakref pool, we should rename this package.

SimonRichardson replied:
Potentially just calling it pool(s)?

arnodel replied:
Technically it's a weakref pool that won't give you weakrefs :). But yeah, there could be a better name.  I think pool/pools is a bit generic, will try to think about a good one.

arnodel replied:
I've renamed it to `luagc`.

SimonRichardson commented on runtime/runtimecontextmanager.go:
```diff
@@ -118,12 +128,27 @@ func (m *runtimeContextManager) PushContext(ctx RuntimeContextDef) {
 	m.messageHandler = ctx.MessageHandler
 	m.parent = &parent
+	if ctx.GCPolicy == IsolateGCPolicy || ctx.HardLimits.Millis > 0 || ctx.HardLimits.Cpu > 0 || ctx.HardLimits.Memory > 0 {
+		m.weakRefPool = weakref.NewDefaultPool()
```
What happens if you already have a pool, do you close that pool before pushing a context? See initRoot, which sets up a pool.

arnodel replied:
If you already have a pool, then you don't close it but it becomes "inactive" until the new context is popped.

SimonRichardson commented on runtime/value.go:
```diff
@@ -76,6 +76,10 @@ func ifaceType(iface interface{}) uintptr {
 	return *(*uintptr)(unsafe.Pointer(&iface))
 }
 
+func ifacePtr(iface interface{}) uintptr {
```
This is horrid :|

arnodel replied:
In its defence I benchmarked this - it made a significant speed difference.  It's only used in Value.Equals() and Value.IsNaN()

arnodel committed:
Update runtime/internal/weakref/clonepool.go

Co-authored-by: Simon Richardson <stickupkid@gmail.com>

arnodel committed:
Rename internal weakref package to luagc

It is not just about managing weakrefs but also about managing finalization
of lua values.

arnodel committed:
Skip runChunk test for noquotas build
